Make compatible with thumbv6m-none-eabi

BjornTheProgrammer opened the pull request:
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Problem with no_std on thumbv6m-none-eabi #302 
Inference on an embedded MCU (RP2040 / Raspberry Pico) #1067
[WIP] Use portable atomics to build for thumbv6m-none-eabi target #374

### Changes

The goal is to get `burn` working on a `thumbv6m-none-eabi` target, which std lib does not have access to `Arc`, and some atomics.

It fixes issues surrounding compilation using a new version of ndarray and exposes some feature flags. It also provides an example model that runs on a `thumbv6m-none-eabi` target.

### Testing

Haven't been able to test completely yet, due to `cubecl` and `cubecl-common` dependencies breaking the build.

BjornTheProgrammer commented:
Checklist of things that need to be done
- [ ] pull request to cubecl and cubecl-common to fix errors with alloc
```
use alloc::{boxed::Box, sync::Arc, task::Wake, vec::Vec};
    |                         ^^^^       ^^^^^^^^^^ no `Wake` in `task`
    |                         |
    |                         could not find `sync` in `alloc`
```
- [ ] Fix testing, so that `./run-checks all` uses the right target
- [ ] Remove all deprecated methods from `ndarray` being used
- [ ] Test on actual hardware

antimora commented on examples/onnx-inference-rp2040/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2023 Mike Panetta
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
```
Great that you are tackling this issue. I skimmed through quickly and I had a quick comment.

The code need to be licensed Apache and MIT just like other examples in this repo. 

You can add your name as an author in the Cargo.toml.   

If you copied code from somewhere, you should attribute this in this notices file: https://github.com/tracel-ai/burn/blob/main/NOTICES.md

antimora commented:
@BjornTheProgrammer FYI, it's sufficient to fix make this work with crates in this test: https://github.com/tracel-ai/burn/tree/main/crates/burn-no-std-tests

BjornTheProgrammer commented:
@antimora I've tested my patch with the example I've constructed and got it working. I've also verified that the results are identical to those of an `std` environment with alloc. However I'm having a single issue, as seen above, would you be willing to investigate it, or have any special insight into why it doesn't work when my version of `cubecl` isn't used?

Here is the error messages I get
```
cargo build
    Updating git repository `https://github.com/tracel-ai/cubecl`
     Locking 8 packages to latest compatible versions
      Adding cubecl v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
      Adding cubecl-common v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
      Adding cubecl-core v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
      Adding cubecl-cuda v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
      Adding cubecl-linalg v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
      Adding cubecl-macros v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
      Adding cubecl-runtime v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
      Adding cubecl-wgpu v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
   Compiling cubecl-common v0.1.1 (https://github.com/tracel-ai/cubecl?rev=dd274b6515f853ab1c82db410d6fc7697060f69e#dd274b65)
   Compiling spin v0.9.8
error[E0599]: no method named `compare_exchange_weak` found for struct `AtomicBool` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/mutex/spin.rs:182:14
    |
180 |   ...   while self
    |  _____________-
181 | | ...       .lock
182 | | ...       .compare_exchange_weak(false, true, Ordering::Acquire, Orde...
    | |           -^^^^^^^^^^^^^^^^^^^^^ method not found in `AtomicBool`
    | |___________|
    |

error[E0599]: no method named `compare_exchange` found for struct `AtomicBool` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/mutex/spin.rs:242:14
    |
240 |   ...   if self
    |  __________-
241 | | ...       .lock
242 | | ...       .compare_exchange(false, true, Ordering::Acquire, Ordering:...
    | |           -^^^^^^^^^^^^^^^^ method not found in `AtomicBool`
    | |___________|
    |

error[E0599]: no method named `compare_exchange` found for struct `core::sync::atomic::AtomicU8` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/once.rs:114:18
    |
112 |   ...   match self
    |  _____________-
113 | | ...       .0
114 | | ...       .compare_exchange(old as u8, new as u8, success, failure)
    | |           -^^^^^^^^^^^^^^^^ method not found in `AtomicU8`
    | |___________|
    |

error[E0599]: no method named `fetch_add` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:265:31
    |
265 |         let value = self.lock.fetch_add(READER, Ordering::Acquire);
    |                               ^^^^^^^^^ method not found in `AtomicUsize`

error[E0599]: no method named `fetch_sub` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:268:23
    |
268 |             self.lock.fetch_sub(READER, Ordering::Relaxed);
    |                       ^^^^^^^^^ method not found in `AtomicUsize`

error[E0599]: no method named `fetch_sub` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:305:23
    |
305 |             self.lock.fetch_sub(READER, Ordering::Release);
    |                       ^^^^^^^^^ method not found in `AtomicUsize`

error[E0599]: no method named `fetch_sub` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:349:19
    |
349 |         self.lock.fetch_sub(READER, Ordering::Release);
    |                   ^^^^^^^^^ method not found in `AtomicUsize`

error[E0599]: no method named `fetch_and` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:363:19
    |
363 |         self.lock.fetch_and(!(WRITER | UPGRADED), Ordering::Release);
    |                   ^^^^^^^^^ method not found in `AtomicUsize`

error[E0599]: no method named `fetch_or` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:415:22
    |
415 | ...   if self.lock.fetch_or(UPGRADED, Ordering::Acquire) & (WRITER | UP...
    |                    ^^^^^^^^ method not found in `AtomicUsize`

error[E0599]: no method named `fetch_sub` found for reference `&'rwlock AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:760:19
    |
760 |         self.lock.fetch_sub(READER, Ordering::Release);
    |                   ^^^^^^^^^ method not found in `&AtomicUsize`

error[E0599]: no method named `fetch_sub` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:770:25
    |
770 |         self.inner.lock.fetch_sub(UPGRADED, Ordering::AcqRel);
    |                         ^^^^^^^^^ method not found in `AtomicUsize`

error[E0599]: no method named `fetch_and` found for struct `AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:782:14
    |
780 | /         self.inner
781 | |             .lock
782 | |             .fetch_and(!(WRITER | UPGRADED), Ordering::Release);
    | |             -^^^^^^^^^ method not found in `AtomicUsize`
    | |_____________|
    |

error[E0599]: no method named `compare_exchange` found for reference `&AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:796:16
    |
796 |         atomic.compare_exchange(current, new, success, failure)
    |                ^^^^^^^^^^^^^^^^ method not found in `&AtomicUsize`

error[E0599]: no method named `compare_exchange_weak` found for reference `&AtomicUsize` in the current scope
   --> /Users/bjorn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/spin-0.9.8/src/rwlock.rs:798:16
    |
798 |         atomic.compare_exchange_weak(current, new, success, failure)
    |                ^^^^^^^^^^^^^^^^^^^^^ method not found in `&AtomicUsize`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `spin` (lib) due to 14 previous errors
warning: build failed, waiting for other jobs to finish...
```

These are just a bunch of errors due to spin not having portable-atomic, but it does? So I'm really confused here.

To reproduce the issue use the following commands
1. `rustup target add thumbv6m-none-eabi`
2. `cd examples/onnx-inference-rp2040`
3. `cargo build`

BjornTheProgrammer commented:
I believe I have actually found the issue.

`burn-core` says this 
```toml
[dependencies]

# ** Please make sure all dependencies support no_std when std is disabled **

burn-common = { path = "../burn-common", version = "0.14.0", default-features = false }
burn-dataset = { path = "../burn-dataset", version = "0.14.0", optional = true, default-features = false }
burn-derive = { path = "../burn-derive", version = "0.14.0" }
burn-tensor = { path = "../burn-tensor", version = "0.14.0", default-features = false }

# ...
```

But `burn-common` violates this agreement, since it has a non-optional dependency `cubecl-common = { workspace = true }`, which is not no_std. Which then causes those errors

antimora commented:
Tagging @nathanielsimard @laggui @syl20bnr to suggest dependency consistency. 

I will take a look closer tomorrow.

BjornTheProgrammer commented:
This pull request only supports `ndarray`. Do you all think that exploration into other backend supports should be done, like `autodiff` or `candle`?

antimora commented:
> This pull request only supports `ndarray`. Do you all think that exploration into other backend supports should be done, like `autodiff` or `candle`?
We mainly target inference. 

We do not have no-std tests for candle. It would be great if candle worked because it's CPU implentation is more optimized than ndarray.

Maybe you could tackle them in a different pr

BjornTheProgrammer commented:
> Tagging @nathanielsimard @laggui @syl20bnr to suggest dependency consistency. 
> 
> I will take a look closer tomorrow. 

Any resolution on this front?

antimora commented:
> > Tagging @nathanielsimard @laggui @syl20bnr to suggest dependency consistency.
> > I will take a look closer tomorrow.
> 
> Any resolution on this front?

Yes, the cubecl-common has std flag enabled by default. You need to disable default features like this:

<img width="1301" alt="image" src="https://github.com/user-attachments/assets/e8e69438-3bbb-455f-89d6-adec0bf1ff15">


I resolved the conflicts of your PR locally and built and tested `burn-no-std-tests` successfully.

BjornTheProgrammer commented:
Based on my testing, this doesn't work. When you try to build for the `thumbv6m-none-eabi` target, you still get many `atomic` and `arc` errors.

To test yourself use the following command in `burn-no-std-tests`.
`RUSTFLAGS="--cfg portable_atomic_unsafe_assume_single_core" cargo build --target thumbv6m-none-eabi`

`thumbv6m-none-eabi` does not support atomic ptrs, despite `thumbv7m-none-eabi` having them. This is the fundamental issue here. Pull request [#52](https://github.com/tracel-ai/cubecl/pull/52) fixes that issue.

BjornTheProgrammer commented on crates/burn-ndarray/src/ops/base.rs:
```diff
@@ -612,7 +612,9 @@ fn arg<E: NdArrayElement, const D: usize>(
-    let output = output.into_shape(Dim(reshape.as_slice())).unwrap();
+    let output = output
+        .into_shape_with_order(Dim(reshape.as_slice()))
+        .unwrap();
```
I'm unsure if this is the proper way of doing this.

BjornTheProgrammer commented on crates/burn-ndarray/src/ops/conv.rs:
```diff
@@ -209,7 +209,7 @@ pub(crate) fn conv2d<E: FloatNdArrayElement, Q: QuantElement>(
 
     let output = output
-        .into_shape([batch_size, out_channels, out_height, out_width])
+        .into_shape_with_order([batch_size, out_channels, out_height, out_width])
```
I'm unsure if this is the proper way of doing this.

BjornTheProgrammer commented on crates/burn-ndarray/src/ops/conv.rs:
```diff
@@ -437,7 +437,7 @@ pub(crate) fn conv3d<E: FloatNdArrayElement, Q: QuantElement>(
 
     let output = output
-        .into_shape([batch_size, out_channels, out_depth, out_height, out_width])
+        .into_shape_with_order([batch_size, out_channels, out_depth, out_height, out_width])
```
I'm unsure if this is the proper way of doing this.

BjornTheProgrammer reviewed (commented):
For the following, I do not know which method is best used instead of the deprecated method. Your input would be appreciated.

BjornTheProgrammer commented:
I've added documentation to the burn book (it should be rewritten more to your standards) and replaced the deprecated methods, I'm unsure if I did so correctly. One of the tests still fails from asserts with `ndarray`. However, it runs on the Raspberry Pi Pico just fine. 

These are the final things that need to be fixed before this can be merged. @antimora would you mind tackling these?

antimora commented:
> I've added documentation to the burn book (it should be rewritten more to your standards) and replaced the deprecated methods, I'm unsure if I did so correctly. One of the tests still fails from asserts with `ndarray`. However, it runs on the Raspberry Pi Pico just fine. 
> 
> 
> 
> These are the final things that need to be fixed before this can be merged. @antimora would you mind tackling these?

I will try. I have been feeling sick lately, so my availability might be sporadic.

antimora commented:
@BjornTheProgrammer is this error you're gettin when building `run-checks.sh no-std`?

```
error: dependents require atomic CAS but not available on this target by default;
       consider enabling one of the `unsafe-assume-single-core` or `critical-section` Cargo features.
       see <https://docs.rs/portable-atomic/latest/portable_atomic/#optional-features> for more.
   --> /Users/dilshod/.cargo/registry/src/index.crates.io-6f17d22bba15001f/portable-atomic-1.7.0/src/lib.rs:424:1
    |
424 | / compile_error!(
425 | |     "dependents require atomic CAS but not available on this target by default;\n\
426 | |     consider enabling one of the `unsafe-assume-single-core` or `critical-section` Cargo features.\n\
427 | |     see <https://docs.rs/portable-atomic/latest/portable_atomic/#optional-features> for more."
428 | | );
    | |_^

error: could not compile `portable-atomic` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
[burn]%
```

antimora commented:
I resolved conflicts and rebase. Let me know what's blocking you.

BjornTheProgrammer commented:
I fixed that error, just by updating the revision of `cubecl`. This is the error I'm getting that I don't know how to fix.

Upon fixing this, the PR should be ready to merge. I'll open this to no longer be a draft.

```
---- ops::base::tests::should_generate_row_major_layout_for_cat stdout ----
thread 'ops::base::tests::should_generate_row_major_layout_for_cat' panicked at crates/burn-ndarray/src/ops/base.rs:653:9:
assertion failed: array.array.is_standard_layout()
```

BjornTheProgrammer commented:
@antimora Fixed all issues, `./run-checks.sh all` works as expected. I have tested on actual hardware. This pull request should be ready to merge.

antimora commented:
@BjornTheProgrammer i will review in coming days. I also others to look into this. This will be a very useful feature especially with your example and a book section!

laggui reviewed (commented):
This looks awesome! Thanks for the great addition including both an example and step-by-step guide in the book 🙂 

I'll make time to review this in the next few days 🙏

antimora reviewed (approved):
LGTM

Thank you for you addition. This feature has been asked a few times. I am glad we support CPUs without atomic instructions.

BjornTheProgrammer commented:
@antimora @laggui I've rebased the commit to the latest on the main, I've also renamed and reordered the commits. Nothing has changed in the content.

antimora commented:
@laggui I hope you'd be able to review it soon so we can merge it before other deviations appear on the main.

laggui commented:
Wow I reviewed this earlier this afternoon, I guess I forgot to submit 😅

laggui reviewed (approved):
Awesome work!

Everything looks good, I just have a minor request:

Can you add a small README to the example, and perhaps rename it to something more explicit instead of rp2040. know it is for the target embedded system, but perhaps something like `raspberry-pi-micro` would be better?

I'll let you decide what you think fits best 😄

BjornTheProgrammer commented:
@laggui I've made the requested changes! Good catch, I didn't think of that at all!

laggui commented:
Thanks for the quick change! Looks good to go 👌🏻

BjornTheProgrammer commented:
Thank you so much everyone. I really appreciate your project! It's bringing forward the possibilities of ai inference!
